HIP/hip-1028.md

```diff
@@ -0,0 +1,157 @@
+---
+hip: 1028
+title: Metadata management via SmartContracts
+author: Mustafa Uzun <mustafa.uzun@limechain.tech>, Luke Lee <luke.lee@swirldslabs.com>
+working-group: Valentin Valkanov <valentin.valkanov@limechain.tech>, David Bakin <david.bakin@swirldsbals.com>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+last-call-date-time:
+created: 2024-08-15
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/pull/1028
+requires: 646, 657, 765
+updated: 2024-08-19
+requested-by: NA
+---
+
+## Abstract
+
+HIP-646/657/765 introduces a new metadata field for Fungible Tokens (FT), Non-Fungible Tokens (NFT) and METADATA key  
+for updating metadata. However, these features are not supported by smart contracts.  
+This proposal aims to enhance Hedera Smart Contract Service (HSCS) by exposing HIP-646/657/765 features.
+
+## Motivation
+
+A user may need to create, update and get info for Fungible Tokens (FT), Non-Fungible Tokens (NFT) and METADATA key through contract.  
+The current version of Hedera Smart Contract Service (HSCS) does not cover these use cases.
+
+## Rationale
+
+The HSCS already has the functionality to create and update tokens, so we would want to extend these.  
+
+However, we will add new functions for getting info for tokens to not break the older version of ``HederaTokenService.sol``  
```

Comment:
These two lines don't "read" well together - first suggests extending the functionality but "functionality" could be confused with meaning the actual methods.  Second line does say new methods will be added.

Some kind of rewording might help.

Let's see ... just an idea ...

```
The HAPI calls for creating tokens and updating token information were modified 
to include the new metadata and metadata key fields in their existing protobuf 
structures - a backward compatible change allowed by protobufs.  But the 
Solidity language doesn't work that way: structs can't inherit from other structs, 
and changing a struct changes the method signature and thus the method selector.

Instead of replacing the existing methods this proposal adds _overloads_ for the
setting methods - methods with the same name but different signatures.  Because
the return type of a Solidity method does _not_ change the signature, the get info
methods have different names.
```

But this is just a suggestion of what could be said (and I'm not very happy with it myself).